File: leet_easy_c++/leet_155/leet_155b.cpp
class MinStack {
public:
    /** initialize your data structure here. */
    MinStack() {
       min.push(INT_MAX);
    }
    
    void push(int x) {
        val.push(x);
        if(x <= min.top()) min.push(x);
    }
    
    void pop() {
        if(val.top() == min.top()) min.pop();
        val.pop();
    }
    
    int top() {
        return val.top();
    }
    
    int getMin() {     
        return min.top();
    }
private:
    stack<int> val;
    stack<int> min;
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

// NOTE: this not nearly as fast as the first implemented one, but it is MUCH simpler.

File: leet_easy_python/leet_155/leet_155b.py
class MinStack(object):

    def __init__(self):
        self.stk = [[0,float('inf')]]
        """
        initialize your data structure here.
        """
        

    def push(self, x):
        last = self.stk[-1][1]
        if x <= last: self.stk.append([x,x])
        else: self.stk.append([x,last])   
        """
        :type x: int
        :rtype: void
        """
        

    def pop(self):
        return self.stk.pop()[0]
        """
        :rtype: void
        """
        

    def top(self):
        return self.stk[-1][0] 
        """
        :rtype: int
        """
        

    def getMin(self):
        return self.stk[-1][1]
        """
        :rtype: int
        """
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()